paddle/fluid/framework/new_executor/interpreter/dependency_builder.h

```diff
@@ -37,17 +23,44 @@ namespace paddle {
 namespace framework {
 namespace interpreter {
 
-// equivalent to add_reader_dependency_pass
-void AddDependencyForReadOp(
-    const std::vector<Instruction>& vec_instruction,
-    std::map<int, std::list<int>>* downstream_map,
-    const std::vector<std::vector<bool>>* op_happens_before = nullptr);
-
-void AddDownstreamOp(
-    int prior_op_idx,
-    int posterior_op_idx,
-    std::map<int, std::list<int>>* op_downstream_map,
-    const std::vector<std::vector<bool>>* op_happens_before = nullptr);
+// DependencyBuilder provides some dependency adding function to handle the
+// dependency that cannot be explicitly expresed by a Program. It is a
+// compromise of the incomplete expression ability of the Program. Do not add
+// too many functions here at will, that will bring great burden to the
+// Interpretercore.
+
+class DependencyBuilder {
+ public:
+  explicit DependencyBuilder() : instructions_(nullptr) {}
+
+  const std::map<int, std::set<int>>& Build(
+      const std::vector<Instruction>& instructions);
+
+  bool OpHappensBefore(int prior_op_idx, int posterior_op_idx);
+
+ private:
+  void AddDependencyForCoalesceTensorOp();
+  void AddDependencyForCommunicationOp();
+  void AddDependencyForRandomOp();
+  void AddDependencyForReadOp();
+  void AddDependencyForSequentialRun();
+
+  void AddDownstreamOp(int prior_op_idx, int posterior_op_idx);
+
+  void BuildDownstreamMap();
+
+  void BuildOpHappensBefore();
+
+  void ShrinkDownstreamMap();
+
+  bool is_build_{false};
+  const std::vector<Instruction>* instructions_;  // not_own
+  size_t op_num_;
+
+  // op_happens_before_[i][j] == true means op[i] happens before op[j]
+  std::vector<std::vector<bool>> op_happens_before_;
+  std::map<int, std::set<int>> op_downstream_map_;
```

Comment:
Explain the meaning of `op_downstream_map_[i] = {a, b, c}` as above

Reply:
Done